src/realm/collection.hpp

```diff
@@ -3,12 +3,71 @@
 
 #include <realm/obj.hpp>
 #include <realm/bplustree.hpp>
+#include <realm/obj_list.hpp>
+#include <realm/table.hpp>
 
 #include <iosfwd>      // std::ostream
 #include <type_traits> // std::void_t
 
 namespace realm {
 
+template <class L>
+struct CollectionIterator;
+
+class CollectionBase {
+public:
+    virtual ~CollectionBase() {}
+
+    /*
+     * Operations that makes sense without knowing the specific type
+     * can be made virtual.
+     */
+    virtual size_t size() const = 0;
+    virtual bool is_null(size_t ndx) const = 0;
+    virtual Mixed get_any(size_t ndx) const = 0;
+    virtual void clear() = 0;
+
+    virtual Mixed min(size_t* return_ndx = nullptr) const = 0;
+    virtual Mixed max(size_t* return_ndx = nullptr) const = 0;
+    virtual Mixed sum(size_t* return_cnt = nullptr) const = 0;
+    virtual Mixed avg(size_t* return_cnt = nullptr) const = 0;
+    virtual std::unique_ptr<CollectionBase> clone_collection() const = 0;
+
+    virtual TableRef get_target_table() const = 0;
+
+    // Modifies a vector of indices so that they refer to values sorted according
+    // to the specified sort order
+    virtual void sort(std::vector<size_t>& indices, bool ascending = true) const = 0;
+    // Modifies a vector of indices so that they refer to distinct values.
+    // If 'sort_order' is supplied, the indices will refer to values in sort order,
+    // otherwise the indices will be in original order.
+    virtual void distinct(std::vector<size_t>& indices, util::Optional<bool> sort_order = util::none) const = 0;
+
+    bool is_empty() const
+    {
+        return size() == 0;
+    }
+
+    virtual const Obj& get_obj() const noexcept = 0;
+    virtual ObjKey get_key() const = 0;
+    virtual bool is_attached() const = 0;
+    virtual bool has_changed() const = 0;
+    virtual ConstTableRef get_table() const noexcept = 0;
+    virtual ColKey get_col_key() const noexcept = 0;
+
+protected:
+    friend class Transaction;
+    CollectionBase() noexcept = default;
+    CollectionBase(const CollectionBase&) noexcept = default;
+    CollectionBase(CollectionBase&&) noexcept = default;
+    CollectionBase& operator=(const CollectionBase&) noexcept = default;
+    CollectionBase& operator=(CollectionBase&&) noexcept = default;
+
+    virtual bool init_from_parent() const = 0;
```

Comment:
It's because they are used from other const methods to refresh the inner `mutable` state of the collection, e.g. in a function such as `size()` or `get()`.